docs/G--Model-System/tree-traversal.md

```diff
@@ -0,0 +1,90 @@
+# Model Object Tree Traversal
+Event Espresso model objects (representing database rows) can depend on other model objects through their foreign keys.
+Those model objects can in turn depend on others. This can make a fairly wide tree of dependencies. This articles documents
+some classes that help traversing that tree, whose first application is performing "cascading deletions" (instead of just
+deleting a single model object, also recursively delete all its dependent model objects.)
+
+First off, discovering all the model objects that depend on a particular model object can take quite a few database queries,
+so it's often best to perform these as part of a batch process.
+
+These classes are in the namespace (and corresponding folder) `\EventEspresso\core\services\orm\tree_traversal\`
+
+## ModelObjNode
+This class wraps a single model object, and contains the logic for discovering its dependent model objects. It can do this
+all at once, or across multiple method invocations. Eg
+
+```php
+// Given a transaction model object...
+$t = EEM_Transaction::instance()->get_one();
+
+// Create a ModelObjNode around it.
+$transaction_node = new ModelObjNode($t->ID(), $t->get_model());
+
+// Then request it discover 10 of its dependent model objects. If we wanted to get all of them at once, we could pass INF constant.
+$transaction_node->visit(10);
+
+// And see what they are (the returned array is indexed first by model name whose value is an array of IDs.)
+$model_names_and_ids = $transaction_node->getIds();
+
+// You can do whatever you want with them, like display them, or delete them. 
+// Let's delete them.
+foreach($model_names_and_ids as $model_name => $ids){
+    $model = EE_Registry::instance()->load_model($model_name); 
+    $model->delete_permanently(
+        [
+            [
+                $model->primary_key_name() => ['IN', $ids]
+            ]
+        ],
+        false // Don't block if there are dependent model objects. They're all on the chopping block!
+    );
+}
+```
+
+Also, `ModelObjNode`s are meant to be easily serializable, so the following code is fine:
+```php
+// Given a transaction model object...
+$t = EEM_Transaction::instance()->get_one();
+
+// Create a ModelObjNode around it.
+$transaction_node = new ModelObjNode($t->ID(), $t->get_model());
+
+// Then request it discover 10 of its dependent model objects. If we wanted to get all of them at once, we could pass INF constant.
+$transaction_node->visit(10);
+
+// Serialize the transaction node.
+$t_serialized = serialize($transaction_node);
+
+// Unserialize it.
+$transaction_node_unserialized = unserialize($t_serialized);
+
+// Keep using it. It will pick up where it left off exploring the tree
+$transaction_node_unserialized->visit(20);
```

Comment:
is this supposed to indicate that you are getting 20 additional rows or to fetch a **total** of 20 rows which includes the first 10?

If you mean the former (which I think you do), then it may be potentially less confusing to keep the number of rows visited the same, but change the comment to something like:

```php
// Keep using it to grab the next 10 rows (ie: 11-20).
// It will pick up where it left off exploring the tree
$transaction_node_unserialized->visit(10);
```

Reply:
ya, it's the former. I'll add your suggestion.